juju/juju
# [JUJU-2160] Tidy up Getdeployer on client

This change tries to improve the client when determining what is being deployed (i.e. `local charm?`, `repository bundle?`, etc.)

The particular goals here are:
- [x] Reuse data gathered in deployment type determination where possible.
- [x] Reduce call count to resolvecharm.
- [ ] Find repository charms first, rather than last.


## Checklist

*If an item is not applicable, use `~strikethrough~`.*


## QA steps

There shouldn't be any change to behavior at all. So all the deploy integration tests need to pass as they are, and the code should be manually checked against redundancies etc.

## Notes & Discussion

There's a tiny little more to do here:
- The tests need to be passing
- Repository charms should be checked first (I saw that item after the fact), but with they way the validation code is structured this should be trivial.
